Inline Base plugin into core classes, extract them to separate files

printercu opened the pull request:
Closes #316

jrochkind commented:
I like it! 

I think it will make it easier to understand the different participating shrine objects. 

On the minus side, when it's not always clear what object a method is defined on, I can't just "find" in the main shrine.rb file to find it.  But I think it's a net plus. 

I wonder if some inline comments around the `class << self` stuff would be useful, like `# so methods can be easily overridden by plugins` above each one.

jrochkind commented:
Oh right, I missed what was going on. 

You hypothetically could fake it the way I mistakenly thought it was being faked:

```ruby
class UploadedFile
   include Module.new do
        def instance_method
```

Or it could be in a ShrineBase class, and `Shrine = Class.new(ShrineBase)`. 

I have no opinion on if either of these kinds of contortions are worth it.

janko-m commented:
@jrochkind That would create anonymous modules in the ancestry tree, which is not ideal for introspection:

```rb
module NamedModule
end

class Class1
  include NamedModule
end

Class1.ancestors #=> [Class1, NamedModule, ...]
```
```rb
class Class2
  include Module.new
end

Class2.ancestors #=> [Class2, #<Module:0x00007fed45826d80>, ...]
```

I think for now it would be safest to have this PR still define the same modules:

* `lib/shrine.rb` – `Shrine`, `Shrine::Plugins::Base::ClassMethods`, `Shrine::Plugins::Base::InstanceMethods`
* `lib/shrine/uploaded_file.rb` – `Shrine::UploadedFile`, `Shrine::Plugins::Base::FileClassMethods`, `Shrine::Plugins::Base::FileMethods`
* `lib/shrine/attacher.rb` – `Shrine::Attacher`, `Shrine::Plugins::Base::AttacherClassMethods`, `Shrine::Plugins::Base::AttacherMethods`
* `lib/shrine/attachment.rb` – `Shrine::Attachment`, `Shrine::Plugins::Base::AttachmentClassMethods`, `Shrine::Plugins::Base::AttachmentMethods`
* `lib/shrine/plugins.rb` – `Shrine::Plugins`

printercu commented:
I see the difference now. There should be test for this to prevent such mistakes in future.

Indeed I don't like the idea of having `Shrine::Plugins::Base::FileClassMethods` defined in `shrine/uploaded_file.rb`, because it's even less expected place to find this module and does not follow ruby's convention on naming files for modules.

I have 3 ideas
- In every `shrine.rb`, `shrine/attachment.rb`, etc. define `ClassMethods` and `InstanceMethods` and include/extend-with them right after definition.
- Define `*Base` supperclasses for Attacher, Attachment, etc. However, it still requires modules for Shrine itself.
- Change `.plugin` method to support `AttacherMethods::ClassMethods` along with `AttacherClassMethods` (for backward compatibility), and make it accept modules along with symbols. So it'll be possible to have `shrine/attacher_methods.rb`, etc. and use `Shrine.plugin Shrine` to include them. This way single file would contain both instance- and class-methods for each class, also following ruby's naming conventions. (For Shrine itself use `UploaderMethods` module).

First I preferred second option as it looks more idiomatic. But it looks like 3d is a better one, also bringing improvements for plugins system.

janko-m commented:
I really like the **1st** option. With it we could have class declaration and its implementation defined in the same place (`lib/shrine.rb`, `lib/shrine/uploaded_file.rb`, ...). I think it would be a bit disappointing to the reader to go to `lib/shrine/uploaded_file.rb` too look for `Shrine::UploadedFile` implementation, only to see that it's empty and they have to go to `lib/shrine/file_methods.rb` instead. Having so many files could in general make navigation harder, I wouldn't like to have more than the ones I mentioned in my previous comment.

> Indeed I don't like the idea of having Shrine::Plugins::Base::FileClassMethods defined in shrine/uploaded_file.rb, because it's even less expected place to find this module and does not follow ruby's convention on naming files for modules.

Yeah, now that I think about it more, I agree.

> and make it accept modules along with symbols

This is already possible, that's how Shrine is currently able to load the `Base` plugin.

> There should be test for this to prevent such mistakes in future.

I was actually surprised there were no test failures. Then I realized it was because all tests work with `Shrine` subclasses, to avoid causing side effects. I don't know how to safely add a test for this, because once one test modifies the `Shrine` class, I don't see a way how it can undo it. Let me know if you have ideas.

janko-m commented:
Ok, I think I also like the 3rd option the best now, as long as modules like `Shrine::FileMethods` are defined together with the class in `lib/shrine/uploaded_file.rb`. And I also like the `Shrine::FileMethods::ClassMethods` idea. I agree that the 1st option would be repetitive with its manual inclusion/extension, it's good that with 3rd option we could still use Shrine's built-in plugin loading.

janko-m commented:
> As of testing, I think it should be possible to define some harmless plugin (maybe just record calls, or enable flags) and apply it to root instance.

Yeah, was thinking about this as well, I will try to write this test.

jrochkind commented:
>  I don't know how to safely add a test for this, because once one test modifies the Shrine class, I don't see a way how it can undo it. Let me know if you have ideas.

Unload and re-load the base Shrine class with `Object.remove_const(:Shrine)`, just for the one test that needs to do this wacky thing?  Just brainstorming.

janko-m commented:
This is the test I came up with: https://github.com/shrinerb/shrine/blob/21d3ee55a127eb2968bae1e4e16d60d7fe6c67f8/test/shrine_test.rb#L408-L451

hmistry commented:
My thoughts:

- I'm looking forward to this file splitting.
- I'd like to see no more than the 5 files Janko listed above. 
- I like our current `#plugin` method which is a lot easier to read and follow what's it's doing - sometimes I find meta programming overcomplicates readability. I'd like new dev's to quickly glance at it and say "ok I follow" and move on the plugin code.
- Yes I too was surprised the tests did not fail. 
- @janko-m I think that test looks good. I did not know of `#owner` but I like it. I personally would also include testing the change in method return value so there's no assumption that because the method is owned by `TestPlugin`, it's return value will be that - this is me being thorough but I'm aware this assumption is almost guaranteed otherwise it'll be a huge Ruby bug or breaking change.

janko-m commented:
> I personally would also include testing the change in method return value so there's no assumption that because the method is owned by TestPlugin, it's return value will be that – this is me being thorough but I'm aware this assumption is almost guaranteed otherwise it'll be a huge Ruby bug or breaking change.

Yeah, you're right, now I mind the indirection of comparing `Method#owner` as well. The reason why I chose not to do it initially is because I didn't feel like properly initializing the core classes. Then I remembered `Class#allocate`, which is makes an instance without calling `#initialize` (`.new` is essentially `.allocate` + `#initialize`), so now it should be proper. https://github.com/shrinerb/shrine/blob/4fdb6b2d252e19981854566c9998397eef4092ec/test/shrine_test.rb#L431-L449

printercu commented:
I've updated PR according to our discussion. In the 2nd commit I've implemented using both `ShrineAttacher` and `Shrine::AttacherMethods` as we discussed. Later I've found that first option (using `Attacher::InstanceMethods`) can also be implemented with plugin system and even makes `.plugin` method simpler. This also removes extra constants from `Shrine` namespace. Please see this in 3rd commit. If it goes, I'll merge them into single one.

janko-m commented:
I don't like how complicated the `Shrine.plugin` method has gotten. I was hoping that the 3rd option would just add 3 lines.

```rb
# ...
self::UploadedFile.extend(plugin::FileMethods::ClassMethods) if defined?(plugin::FileMethods::ClassMethods)
# ...
self::Attacher.extend(plugin::Attacher::ClassMethods) if defined?(plugin::Attacher::ClassMethods)
# ...
self::Attachment.extend(plugin::Attachment::ClassMethods) if defined?(plugin::Attachment::ClassMethods)
# ...
```

I know that technically wouldn't be very DRY, but I like that currently it's clearly spelled out to the reader what gets included into what.

Now that I think about it more, I don't find it ideal that the `Shrine.plugin` method would get bigger just to support loading the base functionality. I think there would be benefit in loading the base functionality separately, because it's not really "just another plugin". I also see your point about adding adding constants to the `Shrine` namespace.

Combining these three thoughts, what do you think about keeping the method module structure that this PR currently has:

```rb
Shrine::InstanceMethods
Shrine::ClassMethods
Shrine::UploadedFile::InstanceMethods
Shrine::UploadedFile::ClassMethods
Shrine::Attacher::InstanceMethods
Shrine::Attacher::ClassMethods
Shrine::Attachment::InstanceMethods
Shrine::Attachment::ClassMethods
```

But instead of including/extending them via `Shrine.plugin`, we include/extend them manually. That's the **1st** option from our discussion. You mentioned that it would be repetitive to have to `include`/`extend` each of them, but what if we did it in a loop at the end of `lib/shrine.rb`:

```rb
# lib/shrine.rb

require "shrine/uploaded_file"
require "shrine/attacher"
require "shrine/attachment"

class Shrine
  # ...
  module ClassMethods
    # ...
  end

  module InstanceMethods
    # ...
  end
  # ...
end

[Shrine, Shrine::UploadedFile, Shrine::Attacher, Shrine::Attachment].each do |core_class|
  core_class.include core_class.const_get(:InstanceMethods)
  core_class.extend core_class.const_get(:ClassMethods)
end
```

What do you think?

Thank you for your continued work on this 🙏

printercu commented:
Sounds reasonable. We can discuss on further `.plugin` improvements later, and keep this PR only for splitting `shrine.rb`.

printercu commented:
Updated. Finally:
- all methods stay untouched, just moved them around,
- slightly updated test which checks that plugin can override core methods,
- updated comments for core classes: changed note about Plugins::Base to note about InstanceMethods and ClassMethods,
- updated website generator.

janko-m commented on lib/shrine/plugins.rb:
```diff
@@ -0,0 +1,29 @@
+  module Plugins
+    @plugins = {}
+
+    module_function
```
Nice addition, thanks.

janko-m commented on lib/shrine.rb:
```diff
@@ -29,966 +31,325 @@ def initialize(io, missing_methods)
+require 'shrine/plugins'
+require 'shrine/attacher'
+require 'shrine/attachment'
+require 'shrine/uploaded_file'
```
Normally I don't mind that files are required elsewhere in the file when needed, but here I don't see any reason why we wouldn't require them on the top. I think it's nice for the reader to see this on the top of the file.

Also, Shrine uses double quotes everywhere, even for `require` statements.

janko-m commented on lib/shrine/uploaded_file.rb:
```diff
@@ -0,0 +1,252 @@
+    end
+
+    module InstanceMethods
+
```
There is an extra newline here

janko-m commented on test/shrine_test.rb:
```diff
@@ -405,57 +405,61 @@ def generate_location(io, context)
-      mod = TestPlugin.const_get(name)
-      mod.send(:undef_method, :foo)
+  describe '.plugin' do
+    describe 'for global plugin' do
```
I think `when called globally` would be more accurate.

Also, the convention is double quotes here as well.

janko-m commented on test/shrine_test.rb:
```diff
@@ -405,57 +405,61 @@ def generate_location(io, context)
+        end
+      end
+
+      it "allows plugins to override base methods of core classes" do
```
`allows the plugin to override base methods of core classes` (`s/plugins/the plugin`)

janko-m commented on www/_data/core.yml:
```diff
@@ -1,21 +1,18 @@
 -
   name: Shrine::Attachment
-  path: Plugins/Base/AttachmentMethods.html
+  path: Shrine/Attachment.html
```
I would prefer if we referenced the `InstanceMethods.html` for these three core classes, because they will hold the meat of the implementation (the `ClassMethods` module will only hold a few methods which I don't think are important to look at).

hmistry commented:
@janko-m It looks good to me now.

@printercu Thank you for working on this and being open to the various feedback. I'm looking forward to working with the separate files. 🙂

janko-m commented:
I've done some minor changes after your commit: https://github.com/shrinerb/shrine/commit/da17e087b5c9040170ac3ffba12fa3ffec792bb9

I reverted the `module_function` addition to `Shrine::Plugins` after all, because for me it makes sense to use it only if `Shrine::Plugins` can be included somewhere else, but I realized that it can't, because it relies on the `@plugins` variable that only exists in the `Shrine::Plugins` module.
